README.md
# fhir-works-on-aws-authz-smart

## Upgrade notice

Versions 3.1.1 and 3.1.2 of the `fhir-works-on-aws-authz-smart` package have been deprecated for necessary security updates. Please upgrade to version 3.1.3 or higher. For more information, see [the fhir-works-on-aws-authz-smart security advisory](https://github.com/awslabs/fhir-works-on-aws-authz-smart/security/advisories/GHSA-vv7x-7w4m-q72f).


## Purpose

This package is an implementation of the authorization interface from the [FHIR Works interface](https://github.com/awslabs/fhir-works-on-aws-interface/blob/mainline/src/authorization.ts). It uses the [Substitutable Medical Applications, Reusable Technologies (SMART on FHIR) specification v1.0.0](http://www.hl7.org/fhir/smart-app-launch/1.0.0) to authorize users. Requests are authorized if the requestor or the patient in context is [**referenced**](https://www.hl7.org/fhir/references.html) in the resource in question.

To use and deploy this component please follow the overall [`smart-mainline` branch README](https://github.com/awslabs/fhir-works-on-aws-deployment/tree/smart-mainline)

## Assumptions

The following assumptions have been made while creating this package:

- An [OAuth2](https://oauth.net/2/) [OpenID Connect](https://openid.net/connect/) authorization server already exists and is used as, or in conjunction with, an identity provider.
  - The OAuth2 server complies with the [SMART on FHIR specification](https://docs.smarthealthit.org/)
  - The OAuth2 server has a JSON Web Key Set endpoint used to get the key for verifying incoming access tokens
- The identity provider has a user claim (either `fhirUser` or `profile`) representing who this user is in context to this FHIR server. This user must be represented by a fully qualified URL in the claim.
  - As an example, the `fhirUser` claim should look like: `https://www.fhir.com/Patient/1234`
  - When using `user` scopes it is assumed that the `fhirUser` will be in the access token to determine who the requestor is
- [`launch` scopes and contextual request](http://www.hl7.org/fhir/smart-app-launch/1.0.0/scopes-and-launch-context/#scopes-for-requesting-context-data) will be handled by the authorization server.
- Once launch context is given to the authorization server it will be included with a `patient` scope and the Patient's resourceType and id in the `launch_response_patient` claim within the access token.
  - As an example, the `launch_response_patient` claim should look like: `Patient/id`

## Authorization

This packages uses SMART scopes and the references found in the resources as a way to determine access. Scopes are used to tell the authorization and resource server what access the requestor has. In addition, the references are used to do further authorization, in an attribute based access control model.

### Scopes

This resource server supports [SMART' v1.0.0 clinical scopes](http://www.hl7.org/fhir/smart-app-launch/1.0.0/scopes-and-launch-context/#scopes-for-requesting-clinical-data). There are some assumptions made on the authorization and resource server relationship:

- For `patient` scopes, there must be a `launch_response_patient` claim in the access token.
- For `user` scopes, there must be a `fhirUser` claim in the access token.
- The access modifiers `read` and `write` will give permissions as defined in the incoming [SMARTConfig](./src/smartConfig.ts).

The resource server also supports [SMART's Flat FHIR or Bulk Data `system` scope](https://hl7.org/fhir/uv/bulkdata/authorization/index.html#scopes). `system` scopes have the format `system/(:resourceType|*).(read|write|*)`– which conveys the same access scope as the matching user format `user/(:resourceType|*).(read|write|*)`. 

### Attribute Based Access Control (ABAC)

This implementation of the SMART on FHIR specification uses attribute based access control. Access to a resource is given if one of the following statements is true:

- The fhirUser making the request is considered an Admin (default configuration makes a Practitioner an admin).
- The fhirUser making the request or the patient in context is looking up their own resource (verified via the `resourceType` and `id`).
- The fhirUser making the request or the patient in context is referenced in the resource in which they are taking action on.

As an example below, the Patient resource is accessible by:

- Admins of the system
- Requests with the usage of the `system` scope
- `Patient/example`: via `resourceType` and `id` check
- `Patient/diffPatient`: because it is referenced in the `link` field
- `Practitioner/DrBell`: because it is referenced in the `generalPractitioner` field

```json
// Example Patient resource with references
{
  "resourceType": "Patient",
  "id": "example",
  "generalPractitioner": [
    {
      "reference": "Practitioner/DrBell"
    }
  ],
  "link": [
    {
      "type": "seealso",
      "other": {
        "reference": "Patient/diffPatient"
      }
    }
  ],
  "address": [
    {
      "period": {
        "start": "1974-12-25"
      },
      "city": "London",
      "use": "home",
      "line": ["221b Baker St"],
      "district": "Marylebone",
      "postalCode": "6XE",
      "text": "221b Baker St, Marylebone, London NW1 6XE, United Kingdom",
      "type": "both"
    }
  ],
  "deceasedBoolean": false,
  "name": [
    {
      "family": "Holmes",
      "given": ["Sherlock"],
      "use": "official"
    }
  ],
  "gender": "male",
  "active": true
}
```

## Usage

Add this package to your `package.json` file and install as a dependency. For usage examples please see the deployment component's [package.json](https://github.com/awslabs/fhir-works-on-aws-deployment/blob/smart-mainline/package.json)

### Configuration

The SMART specification gives a lot of room for interpretation between the resource and authorization server relationship. With this in mind we developed our SMART implementation to be flexible. The configurations currently available can be viewed in the [SMARTConfig](./src/smartConfig.ts).

### SMART on FHIR scope rules

Within the [SMARTConfig](./src/smartConfig.ts) you can see an example implementation of a ScopeRule. The ScopeRule says which operations a scope gives access to. For example, the `user/*.write` scope provides access to 'create' resource but not 'update' resource.

For an example usage of the SMARTConfig, please see [authZConfig.ts](https://github.com/awslabs/fhir-works-on-aws-deployment/blob/smart-mainline/src/authZConfig.ts) in the deployment package.

## Dependency tree

This package is dependent on:

- [interface component](https://github.com/awslabs/fhir-works-on-aws-interface)
  - This package defines the interface we are trying to use

## Known issues

You can track the issues on the GitHub repository.

## Security

See [CONTRIBUTING](CONTRIBUTING.md#security-issue-notifications) for more information.

## License

This project is licensed under the Apache-2.0 License.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [3.1.4](https://github.com/awslabs/fhir-works-on-aws-authz-smart/compare/v3.1.3...v3.1.4) (2023-01-26)

### Tests

* Added more extensive tests ([#105](https://github.com/awslabs/fhir-works-on-aws-authz-smart/pull/105), [#104](https://github.com/awslabs/fhir-works-on-aws-authz-smart/pull/104), [#103](https://github.com/awslabs/fhir-works-on-aws-authz-smart/pull/103), [#102](https://github.com/awslabs/fhir-works-on-aws-authz-smart/pull/102), [#101](https://github.com/awslabs/fhir-works-on-aws-authz-smart/pull/101), [#100](https://github.com/awslabs/fhir-works-on-aws-authz-smart/pull/100), [#99](https://github.com/awslabs/fhir-works-on-aws-authz-smart/pull/99))

### [3.1.3](https://github.com/awslabs/fhir-works-on-aws-authz-smart/compare/v3.1.2...v3.1.3) (2022-09-12)


### Bug Fixes

* handle include/revInclude correctly ([#92](https://github.com/awslabs/fhir-works-on-aws-authz-smart/issues/92)) ([203bbc0](https://github.com/awslabs/fhir-works-on-aws-authz-smart/commit/203bbc0dd17de748c36b33c68b866ed2dfd613d3))

### [3.1.2](https://github.com/awslabs/fhir-works-on-aws-authz-smart/compare/v3.1.1...v3.1.2) (2022-09-08)


### Bug Fixes

* only allow scopes to be returned iff it matches resourceType ([#88](https://github.com/awslabs/fhir-works-on-aws-authz-smart/issues/88)) ([18b059e](https://github.com/awslabs/fhir-works-on-aws-authz-smart/commit/18b059e2eac0cb7583fcc190d00a9c5f555abcdb))

### [3.1.1](https://github.com/awslabs/fhir-works-on-aws-authz-smart/compare/v3.1.0...v3.1.1) (2022-03-11)


### Bug Fixes

* scopes should not be filtered in search use cases ([#79](https://github.com/awslabs/fhir-works-on-aws-authz-smart/issues/79)) ([67af26d](https://github.com/awslabs/fhir-works-on-aws-authz-smart/commit/67af26df03d3c9be9f85e98e8e18230446bde28f))

## [3.1.0](https://github.com/awslabs/fhir-works-on-aws-authz-smart/compare/v3.0.0...v3.1.0) (2021-10-06)


### Features

* add optional headers for JWKS client ([#59](https://github.com/awslabs/fhir-works-on-aws-authz-smart/issues/59)) ([43e1439](https://github.com/awslabs/fhir-works-on-aws-authz-smart/commit/43e14391200a775fd11078cc1de4d23c244074d5))

## [3.0.0](https://github.com/awslabs/fhir-works-on-aws-authz-smart/compare/v2.2.3...v3.0.0) (2021-09-15)


### ⚠ BREAKING CHANGES

* User scope will not be allowed for system export by default. To allow user scope for system export, please update `src/config.ts` in the deployment package to pass in parameter `isUserScopeAllowedForSystemExport`. 

### Features

* update system export to only allow system scope by default ([#60](https://github.com/awslabs/fhir-works-on-aws-authz-smart/issues/60)) ([45b7960](https://github.com/awslabs/fhir-works-on-aws-authz-smart/commit/45b796031c36960b8f5be1b765beee839d0d9aa8))


### Bug Fixes

* remove filtered responses from total count ([#63](https://github.com/awslabs/fhir-works-on-aws-authz-smart/issues/63)) ([82b6ff3](https://github.com/awslabs/fhir-works-on-aws-authz-smart/commit/82b6ff372b44ca40dc3575e7505695f79747d6de))

### [2.2.3](https://github.com/awslabs/fhir-works-on-aws-authz-smart/compare/v2.2.2...v2.2.3) (2021-09-10)


### Bug Fixes

* scope checking for getting status on an export ([#64](https://github.com/awslabs/fhir-works-on-aws-authz-smart/issues/64)) ([69f9297](https://github.com/awslabs/fhir-works-on-aws-authz-smart/commit/69f9297f7721113667f2c8b9ef3c65ac06016c49))

### [2.2.2](https://github.com/awslabs/fhir-works-on-aws-authz-smart/compare/v2.2.1...v2.2.2) (2021-08-23)


### Bug Fixes

* group export permission update ([#56](https://github.com/awslabs/fhir-works-on-aws-authz-smart/issues/56)) ([aff43d7](https://github.com/awslabs/fhir-works-on-aws-authz-smart/commit/aff43d79032e94589ec059f165ab3039b2a48ef3))

### [2.2.1](https://github.com/awslabs/fhir-works-on-aws-authz-smart/compare/v2.2.0...v2.2.1) (2021-08-19)


### Bug Fixes

* pass fhirServiceBaseUrl in bundle request auth ([#53](https://github.com/awslabs/fhir-works-on-aws-authz-smart/issues/53)) ([a87adcf](https://github.com/awslabs/fhir-works-on-aws-authz-smart/commit/a87adcfa38e13c39cb8fd524a12b5b2747f66170))

## [2.2.0](https://github.com/awslabs/fhir-works-on-aws-authz-smart/compare/v2.1.1...v2.2.0) (2021-08-17)


### Features

* use fhirServiceBaseUrl from requests if present ([#50](https://github.com/awslabs/fhir-works-on-aws-authz-smart/issues/50)) ([31f951a](https://github.com/awslabs/fhir-works-on-aws-authz-smart/commit/31f951a660b4f9f4f4ad52b95670063f175868b2)), closes [#45](https://github.com/awslabs/fhir-works-on-aws-authz-smart/issues/45)

### [2.1.1](https://github.com/awslabs/fhir-works-on-aws-authz-smart/compare/v2.1.0...v2.1.1) (2021-08-16)


### Bug Fixes

* Correctly pass params in url for introspect jwt flow ([#48](https://github.com/awslabs/fhir-works-on-aws-authz-smart/issues/48)) ([d67b693](https://github.com/awslabs/fhir-works-on-aws-authz-smart/commit/d67b6932f5a1a0bbe6c340bd340bc2351068adbf))

## [2.1.0](https://github.com/awslabs/fhir-works-on-aws-authz-smart/compare/v2.0.0...v2.1.0) (2021-07-23)


### Features

* allow regex in SMARTConfig.expectedAudValue ([#43](https://github.com/awslabs/fhir-works-on-aws-authz-smart/issues/43)) ([2442856](https://github.com/awslabs/fhir-works-on-aws-authz-smart/commit/244285606d70d7bed1a37a6bc19d32063af036ec))
* allow token introspection as an authz option ([#44](https://github.com/awslabs/fhir-works-on-aws-authz-smart/issues/44)) ([f123621](https://github.com/awslabs/fhir-works-on-aws-authz-smart/commit/f123621a4f65819f89e161ec3fbba1ac46301f41))


### Bug Fixes

* support system scope ([#46](https://github.com/awslabs/fhir-works-on-aws-authz-smart/issues/46)) ([7ae2f49](https://github.com/awslabs/fhir-works-on-aws-authz-smart/commit/7ae2f49cd20e933e45d40b9e518b8ec72c3d61b6))


src/comboTest/testStubs.ts
import { BatchReadWriteRequest } from 'fhir-works-on-aws-interface';
import { FhirResource, ScopeRule, SMARTConfig } from '../smartConfig';
import { getFhirUser } from '../smartAuthorizationHelper';
import { SEARCH_OPERATIONS } from '../smartScopeHelper';

export const scopeRule = (): ScopeRule => ({
    patient: {
        read: ['read', 'vread', 'search-type', 'search-system', 'history-instance', 'history-type', 'history-system'],
        write: ['create', 'transaction'],
    },
    user: {
        read: ['read', 'vread', 'search-type', 'search-system', 'history-instance', 'history-type', 'history-system'],
        write: ['update', 'patch', 'create', 'delete', 'transaction'],
    },
    system: {
        // "read" allows system export and group export
        read: ['read'],
        write: [],
    },
});
export const expectedAud = 'api://default';
export const expectedIss = 'https://dev-6460611.okta.com/oauth2/default';
export const baseAuthZConfig = (): SMARTConfig => ({
    version: 1.0,
    scopeKey: 'scp',
    scopeRule: scopeRule(),
    expectedAudValue: expectedAud,
    expectedIssValue: expectedIss,
    fhirUserClaimPath: 'fhirUser',
    launchContextPathPrefix: 'ext.launch_response_',
    jwksEndpoint: `${expectedIss}/jwks`,
});
export const apiUrl = 'https://fhir.server.com/dev';
export const fakeUrl = 'https://random.server.com/dev';
export const id = 'id';
export const patientId = `Patient/${id}`;
export const practitionerId = `Practitioner/${id}`;
export const patientIdentity = `${apiUrl}/${patientId}`;
export const practitionerIdentity = `${apiUrl}/${practitionerId}`;
export const sub = 'test@test.com';

export const patientContext: any = {
    ext: { launch_response_patient: patientIdentity },
};
export const patientFhirUser: any = {
    fhirUser: patientIdentity,
};
export const practitionerFhirUser: any = {
    fhirUser: practitionerIdentity,
};

export const baseAccessNoScopes: any = {
    ver: 1,
    jti: 'AT.6a7kncTCpu1X9eo2QhH1z_WLUK4TyV43n_9I6kZNwPY',
    iss: expectedIss,
    aud: expectedAud,
    iat: 1668546607,
    exp: 2668546607,
    cid: '0oa8muazKSyk9gP5y5d5',
    uid: '00u85ozwjjWRd17PB5d5',
    sub,
};

const validPatient = {
    resourceType: 'Patient',
    id,
    meta: {
        versionId: '1',
        lastUpdated: '2020-06-28T12:03:29.421+00:00',
    },
    name: [
        {
            given: ['JONNY'],
        },
    ],
    gender: 'male',
    birthDate: '1972-10-13',
    address: [
        {
            city: 'Ruppertshofen',
        },
    ],
};

const validPatientObservation = {
    resourceType: 'Observation',
    id: '1274045',
    meta: {
        versionId: '1',
        lastUpdated: '2020-06-28T12:55:47.134+00:00',
    },
    status: 'final',
    code: {
        coding: [
            {
                system: 'http://loinc.org',
                code: '15074-8',
                display: 'Glucose [Moles/volume] in Blood',
            },
        ],
    },
    subject: {
        reference: patientIdentity,
        display: 'JONNY',
    },
    effectivePeriod: {
        start: '2013-04-02T09:30:10+01:00',
    },
    issued: '2013-04-03T15:30:10+01:00',
    valueQuantity: {
        value: 6.3,
        unit: 'mmol/l',
        system: 'http://unitsofmeasure.org',
        code: 'mmol/L',
    },
    interpretation: [
        {
            coding: [
                {
                    system: 'http://terminology.hl7.org/CodeSystem/v3-ObservationInterpretation',
                    code: 'H',
                    display: 'High',
                },
            ],
        },
    ],
    referenceRange: [
        {
            low: {
                value: 3.1,
                unit: 'mmol/l',
                system: 'http://unitsofmeasure.org',
                code: 'mmol/L',
            },
            high: {
                value: 6.2,
                unit: 'mmol/l',
                system: 'http://unitsofmeasure.org',
                code: 'mmol/L',
            },
        },
    ],
};

const validCondition: any = {
    resourceType: 'Condition',
    id: 'example',
    text: {
        status: 'generated',
        div: '<div xmlns="http://www.w3.org/1999/xhtml">Severe burn of left ear (Date: 24-May 2012)</div>',
    },
    clinicalStatus: {
        coding: [
            {
                system: 'http://terminology.hl7.org/CodeSystem/condition-clinical',
                code: 'active',
            },
        ],
    },
    verificationStatus: {
        coding: [
            {
                system: 'http://terminology.hl7.org/CodeSystem/condition-ver-status',
                code: 'confirmed',
            },
        ],
    },
    category: [
        {
            coding: [
                {
                    system: 'http://terminology.hl7.org/CodeSystem/condition-category',
                    code: 'encounter-diagnosis',
                    display: 'Encounter Diagnosis',
                },
                {
                    system: 'http://snomed.info/sct',
                    code: '439401001',
                    display: 'Diagnosis',
                },
            ],
        },
    ],
    severity: {
        coding: [
            {
                system: 'http://snomed.info/sct',
                code: '24484000',
                display: 'Severe',
            },
        ],
    },
    code: {
        coding: [
            {
                system: 'http://snomed.info/sct',
                code: '39065001',
                display: 'Burn of ear',
            },
        ],
        text: 'Burnt Ear',
    },
    bodySite: [
        {
            coding: [
                {
                    system: 'http://snomed.info/sct',
                    code: '49521004',
                    display: 'Left external ear structure',
                },
            ],
            text: 'Left Ear',
        },
    ],
    subject: {
        reference: patientIdentity,
    },
    onsetDateTime: '2012-05-24',
};

const validMedicationRequest: any = {
    resourceType: 'MedicationRequest',
    id: 'self-tylenol',
    text: {
        status: 'generated',
        div: '<div xmlns="http://www.w3.org/1999/xhtml"><p><b>Generated Narrative</b></p><p><b>id</b>: self-tylenol</p><p><b>identifier</b>: id: 12345689 (OFFICIAL)</p><p><b>status</b>: active</p><p><b>intent</b>: plan</p><p><b>reported</b>: true</p><p><b>medication</b>: <span title="Codes: {http://www.nlm.nih.gov/research/umls/rxnorm 1187314}">Tylenol PM Pill</span></p><p><b>subject</b>: <a href="Patient-example.html">Amy V. Shaw. Generated Summary: id: example; Medical Record Number: 1032702 (USUAL); active; Amy V. Shaw , Amy V. Baxter ; ph: 555-555-5555(HOME), amy.shaw@example.com; gender: female; birthDate: 1987-02-20</a></p><p><b>encounter</b>: <a href="Encounter-example-1.html">Office Visit. Generated Summary: id: example-1; status: finished; <span title="{http://terminology.hl7.org/CodeSystem/v3-ActCode AMB}">ambulatory</span>; <span title="Codes: {http://www.ama-assn.org/go/cpt 99201}">Office Visit</span>; period: 02/11/2015 9:00:14 AM --&gt; 02/11/2015 10:00:14 AM</a></p><p><b>authoredOn</b>: 2019-06-24</p><p><b>requester</b>: <a href="Patient-example.html">**self-prescribed**. Generated Summary: id: example; Medical Record Number: 1032702 (USUAL); active; Amy V. Shaw , Amy V. Baxter ; ph: 555-555-5555(HOME), amy.shaw@example.com; gender: female; birthDate: 1987-02-20</a></p><p><b>dosageInstruction</b>: </p></div>',
    },
    identifier: [{ use: 'official', system: 'http://acme.org/prescriptions', value: '12345689' }],
    status: 'active',
    intent: 'plan',
    reportedBoolean: true,
    medicationCodeableConcept: {
        coding: [
            { system: 'http://www.nlm.nih.gov/research/umls/rxnorm', code: '1187314', display: 'Tylenol PM Pill' },
        ],
        text: 'Tylenol PM Pill',
    },
    subject: { reference: patientIdentity },
    encounter: { reference: 'Encounter/example-1', display: 'Office Visit' },
    authoredOn: '2019-06-24',
    requester: { reference: 'Patient/example', display: '**self-prescribed**' },
    dosageInstruction: [{ text: 'Takes 1-2 tablets once daily at bedtime as needed for restless legs' }],
};

export const getFhirUserType = (fhirUser: string | undefined): string | undefined => {
    if (fhirUser === 'practitionerFhirUser') {
        return practitionerIdentity;
    }
    if (fhirUser === 'patientFhirUser' || fhirUser === 'patientIdentity') {
        return patientIdentity;
    }
    return undefined;
};

export const getFhirUserObject = (fhirUser: string | undefined): FhirResource | undefined => {
    if (fhirUser === 'practitionerFhirUser') {
        return getFhirUser(practitionerIdentity);
    }
    if (fhirUser === 'patientFhirUser' || fhirUser === 'patientIdentity') {
        return getFhirUser(patientIdentity);
    }
    return undefined;
};

export const convertNAtoUndefined = (s: string) => (s === 'N/A' ? undefined : s);

export const convertToBaseUrl = (url: string): string | undefined => {
    if (url === 'matchUrl') {
        return apiUrl;
    }
    if (url === 'nonMatchUrl') {
        return fakeUrl;
    }
    return undefined;
};

export type ResourceBodyDescription =
    | 'matchObservation'
    | 'unmatchCondition'
    | 'matchPatient'
    | 'unmatchPatient'
    | 'patientReferencePractitioner';

export const getResourceBody = (resourceBodyDescription: ResourceBodyDescription) => {
    switch (resourceBodyDescription) {
        case 'matchObservation':
            return validPatientObservation;
        case 'unmatchCondition':
            return {
                ...validCondition,
                subject: {
                    reference: `${apiUrl}/another-id`,
                },
            };
        case 'matchPatient':
            return validPatient;
        case 'unmatchPatient':
            return { ...validPatient, id: 'another-id' };
        case 'patientReferencePractitioner':
            return { ...validPatient, generalPractitioner: { reference: practitionerIdentity } };
        default:
            return undefined;
    }
};

export const getResourceType = (resourceBodyDescription: ResourceBodyDescription | undefined) => {
    if (!resourceBodyDescription) {
        return undefined;
    }
    switch (resourceBodyDescription) {
        case 'matchObservation':
            return 'Observation';
        case 'unmatchCondition':
            return 'Condition';
        default:
            return 'Patient';
    }
};

export const generateSearchBundle = (
    condition?: boolean,
    medicationRequest?: boolean,
    patient?: boolean,
    unmatchPatient?: boolean,
) => {
    const items = [];
    if (condition) {
        items.push({
            resource: { ...validCondition, subject: undefined },
        });
    }
    if (medicationRequest) {
        items.push({
            resource: validMedicationRequest,
        });
    }
    if (patient) {
        items.push({
            resource: validPatient,
        });
    }
    if (unmatchPatient) {
        items.push({
            resource: { ...validPatient, id: 'notSamePatient' },
        });
    }
    return items;
};
export const getReadResponseAndOperation = (
    matchMedicationRequest: boolean,
    matchPatient: boolean,
    unmatchCondition: boolean,
    unmatchPatient: boolean,
) => {
    const searchBundle = generateSearchBundle(unmatchCondition, matchMedicationRequest, matchPatient, unmatchPatient);
    if (searchBundle.length === 1) {
        return {
            readResponse: searchBundle[0],
            operation: 'read',
        };
    }
    return {
        readResponse: {
            total: searchBundle.length,
            entry: searchBundle,
        },
        operation: SEARCH_OPERATIONS[0],
    };
};

export const generateBundle = (): BatchReadWriteRequest[] => {
    return [
        {
            operation: 'create',
            resourceType: 'Observation',
            id: validPatientObservation.id,
            resource: validPatientObservation,
            // references generated as per this method in routing: https://github.com/awslabs/fhir-works-on-aws-routing/blob/mainline/src/router/bundle/bundleParser.ts#L328
            references: [
                {
                    resourceType: 'Patient',
                    id: patientId,
                    vid: '1',
                    rootUrl: apiUrl,
                    referenceFullUrl: patientIdentity,
                    referencePath: 'subject',
                },
            ],
        },
        {
            operation: 'create',
            resourceType: 'Condition',
            id: validCondition.id,
            resource: { ...validCondition, subject: undefined }, // remove reference to patient
        },
        {
            operation: 'read',
            resourceType: 'Patient',
            id,
            resource: undefined,
        },
        {
            operation: 'read',
            resourceType: 'Patient',
            id: 'PatientNotSameId',
            resource: undefined,
        },
    ];
};
